File: 1219. Path with Maximum Gold.py
class Solution:
    def getMaximumGold(self, grid: List[List[int]]) -> int:
        def dfs(curr_x, curr_y):
            nonlocal grid
            temp = grid[curr_x][curr_y]
            answer = 0
            grid[curr_x][curr_y] = 0
            for next_x, next_y in [(curr_x+1, curr_y), (curr_x-1, curr_y), (curr_x, curr_y+1), (curr_x, curr_y-1)]:
                if 0 <= next_x < len(grid) and 0 <= next_y < len(grid[0]) and grid[next_x][next_y] != 0:
                    answer = max(answer, dfs(next_x, next_y))
            grid[curr_x][curr_y] = temp
            return answer + temp
        answer = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] != 0:
                    answer = max(answer, dfs(i, j))
        return answer

File: 1219. Path with Maximum Gold.java
class Solution {
    private static final int[][] directions = {{0,1},{0,-1},{1,0},{-1,0}};
    public int getMaximumGold(int[][] grid) {
        int answer = 0;
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[i].length; j++) {
                if (grid[i][j] != 0) answer = Math.max(answer, dfs(grid, i, j));
            }
        }
        return answer;
    }
    private int dfs(int[][] grid, int currX, int currY) {
        int temp = grid[currX][currY];
        int answer = 0;
        grid[currX][currY] = 0;
        for (int[] d : directions) {
            int nextX = currX + d[0];
            int nextY = currY + d[1];
            if (nextX < 0 || nextX >= grid.length) continue;
            if (nextY < 0 || nextY >= grid[0].length) continue;
            if (grid[nextX][nextY] == 0) continue;
            answer = Math.max(answer, dfs(grid, nextX, nextY));
        }
        grid[currX][currY] = temp;
        return temp + answer;
    }
}